### ISSUE TITLE
Persistent Streams Question: How to make pulling agent read faster?

### ISSUE THREAD

**Issue by amamh** [dotnet/orleans#1448]
I highlighted my questions if you want to skip reading.

In my app, I have a problem because data is being sent over the persistent stream faster than it's being read by the server.

That causes the physical queue in the persistent stream provider to keep growing since it's enqueuing faster than it's dequeuing.

I think the solution is:
- Throttle the "pushing" client.
  - I can do this manually in my code somehow.
  - **Can we have backpressure here in Orleans framework itself?**

- Increase reading speed:
  - Increase the number of physical queues which, I believe, will increase the number of pulling agents and server "AdapterReceiver"'s. It's already equal to the number of cores on my PC, so it doesn't have a positive effect to increase it further.
  - Spawn another server instance on a different PC.
  - **What else can I do to make the server read faster?**

Any help is appreciated. Also, just for reference, [this is my stream provider implementation](https://github.com/amamh/orleans-PipeStreamProvider) which uses Redis as the physical queue.

**Comment by veikkoeeva** [comment]
I'll reference [recent gitter](https://gitter.im/dotnet/orleans?at=56b1ab5a80ad69394a7c17cb) about back pressure. The Rx* people have also debated on adding back pressure to streams and especially RxJava went to do it. My tl;dr interpretation of the debate: not a good idea.

**Comment by amamh** [comment]
@veikkoeeva 

It's honestly a bad idea in my case too because it would mean the clients are receiving old data. I have to make sure that there is no slowing down.

**Comment by jason-bragg** [comment]
@amamh 
Backpressure is not build into the agent, it's currently regulated by the adapter; Primarily the queue cache's IsUnderPressure and MaxAddCount calls.  For real-time processing, throttling is not great, but can be necessary in some cases.

Simplest (not best, just simple) approach is to figure max number of events the service can handle, and have the MaxAddCount vary based off of this value and the time between calls.  Say that the service can process a max of 2000 events per second and the delta in time between the last MaxAddCount call and the current MaxAddCount call was 5ms, then MaxAddCount should return 10 (2000/s = 2/ms,  2/ms*5ms =10).  This will mean the receiver will be asked for up to 10 messages.  If the receiver honors the requested size, the queue will be read at roughly the target rate of ~2000 events per second.

> Spawn another server instance on a different PC

I suggest using implicit streams (with ImplicitOnly pubsub setting) to grains that are local to the agent, either [PreferLocalPlacement] or [StatelessWorker(1)], then add more hardware.

Optimize
What is the expected load on your service?
- What is the target throughput per event queue?
- What is the target throughput per stream?
- How expensive is your event processing?  ticks?  milliseconds? seconds?

How many streams do you roughly expect per queue? This matters for parallelization. If there are 2 streams, then all your events will be processed by 2 grains. If there are 1000 streams, 1000 grains can service the events. For normal (non-reentrant) grains, this effects how much work can be done in parallel.  Too much parallelization can be costly as well; depends on the type of work being done.

It may be valuable for you to run a load test with event processing in your grains turned off to measure the max event delivery rate of your system.  This information can be very valuable for optimization.

**Comment by amamh** [comment]
Hey, thanks for the reply!

> Primarily the queue cache's IsUnderPressure and MaxAddCount calls

That only controls the receiving side, It doesn't affect the writing speed at all. Right?

> implicit streams (with ImplicitOnly pubsub setting) to grains 

Why would implicit stream subscription be different than explicitly subscribing? Also, I'm currently subscribing from "normal" objects not grains. The problem was that the server isn't reading (dequeuing) from the queue fast enough, so the queues were growing in size continuously.

> How many streams do you roughly expect per queue

I set number of queues = number of cores, in my case just 4. I have in total about 20 thousand streams. They get updated (data pushed) every second. However, the case that was failing (queue growing continuously) was a lot higher than that and was due to a mistake in the code. It was the same number of streams but the update rate was something like 10 times more frequent.

------

I have another issue related to this, but I think it should be in a separate thread. I'll give a brief summary here: I'm calling "OnNextAsync" for pushing data from a non-async method, so it's spawning a new thread (I think?) for every call and not waiting for it to finish before calling it again and spawning the next thread. So, when I had 20k streams being updated every 100ms, it was creating huge pressure on the queue. I feel like I'm doing something wrong there.

**Comment by jason-bragg** [comment]
>That only controls the receiving side, It doesn't affect the writing speed at all. Right?

Correct.  Persistent stream provider assumes that queue is persistent, and data can wait there if required, so it throttles read rate to prevent queue cache from blowing up.  I apologize for misunderstanding your initial question.  Your concern is not the queue cache in the adapter, but the physical queue.  We've no mechanism to throttle clients. :/ 

>Why would implicit stream subscription be different than explicitly subscribing?

Implicit subscriptions are faster to setup and tare down streams (subscribe/unsubscribe).  If you've a fixed set of streams this does not matter, but it can effect perf if your streams come and go over time.  

> I have in total about 20 thousand streams. They get updated (data pushed) every second.

So that's about 20k/second events, over 4 partitions?  So that's 5k per second per partition?  That could be rough on a single machine, depending on event processing time.

>I'm currently subscribing from "normal" objects not grains.

I don't know what you mean.  We only support streams from Orleans clients or grains in silos.  The fastest stream processing pattern I'm aware of is stream processing by local grains (as described in previous post) as that pattern minimizes cost of grain calls, but in your case, if you're on a single silo, all grain calls should be local.  How expensive is your event processing?

**Comment by jason-bragg** [comment]
>I'm calling "OnNextAsync" for pushing data from a non-async method.

are you waiting for the task to complete?
o.OnNextAsync(...).Wait().  ?

**Comment by amamh** [comment]
@jason-bragg 
no

**Comment by jason-bragg** [comment]
@amamh 
Yup.  That's bad. If you want to use the asyn calls in a synchronous manner you should wait().
